// @sanity/color-input/lib/ColorPicker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _common = require("react-color/lib/components/common");

var _ColorPickerFields = _interopRequireDefault(require("./ColorPickerFields"));

var _trashIcon = _interopRequireDefault(require("part:@sanity/base/trash-icon"));

var _ColorPicker = _interopRequireDefault(require("./ColorPicker.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ColorPicker = (_ref) => {
  var width = _ref.width,
      rgb = _ref.rgb,
      hex = _ref.hex,
      hsv = _ref.hsv,
      hsl = _ref.hsl,
      onChange = _ref.onChange,
      onUnset = _ref.onUnset,
      disableAlpha = _ref.disableAlpha,
      renderers = _ref.renderers,
      readOnly = _ref.readOnly;
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width
    }
  }, !readOnly && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: _ColorPicker.default.saturation
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _ColorPicker.default.saturationInner
  }, /*#__PURE__*/_react.default.createElement(_common.Saturation, {
    is: "Saturation",
    onChange: onChange,
    hsl: hsl,
    hsv: hsv
  }))), /*#__PURE__*/_react.default.createElement("div", {
    className: _ColorPicker.default.hue
  }, /*#__PURE__*/_react.default.createElement(_common.Hue, {
    is: "Hue",
    hsl: hsl,
    onChange: !readOnly && onChange,
    style: {
      radius: '2px',
      shadow: 'inset 0 0 0 1px rgba(0,0,0,.15), inset 0 0 4px rgba(0,0,0,.25)'
    }
  })), !disableAlpha && /*#__PURE__*/_react.default.createElement("div", {
    className: _ColorPicker.default.alpha
  }, /*#__PURE__*/_react.default.createElement(_common.Alpha, {
    is: "Alpha",
    rgb: rgb,
    hsl: hsl,
    renderers: renderers,
    onChange: onChange,
    style: {
      radius: '2px',
      shadow: 'inset 0 0 0 1px rgba(0,0,0,.15), inset 0 0 4px rgba(0,0,0,.25)'
    }
  }))), /*#__PURE__*/_react.default.createElement("div", {
    className: _ColorPicker.default.controls
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _ColorPicker.default.preview
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _ColorPicker.default.checkboard
  }, /*#__PURE__*/_react.default.createElement(_common.Checkboard, null)), /*#__PURE__*/_react.default.createElement("div", {
    className: _ColorPicker.default.color,
    style: {
      backgroundColor: "rgba(".concat(rgb.r, ",").concat(rgb.g, ",").concat(rgb.b, ",").concat(rgb.a, ")")
    }
  }), readOnly && /*#__PURE__*/_react.default.createElement("div", {
    className: _ColorPicker.default.readOnly
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("h3", null, hex), /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement("strong", null, "RGB"), " ", rgb.r, " ", rgb.g, " ", rgb.b, " \xA0 \xA0", /*#__PURE__*/_react.default.createElement("strong", null, "HSL"), ": ", Math.round(hsl.h), " ", Math.round(hsl.s), "% ", Math.round(hsl.l), "%")), /*#__PURE__*/_react.default.createElement(_default2.default, {
    disabled: true,
    title: "This color can not be changed (read only)"
  }, "Change"))), !readOnly && /*#__PURE__*/_react.default.createElement("div", {
    className: _ColorPicker.default.fields
  }, /*#__PURE__*/_react.default.createElement(_ColorPickerFields.default, {
    rgb: rgb,
    hsl: hsl,
    hex: hex,
    onChange: onChange,
    disableAlpha: disableAlpha
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    onClick: onUnset,
    title: "Delete color",
    icon: _trashIcon.default,
    color: "danger"
  }))));
};

ColorPicker.propTypes = {
  width: _propTypes.default.string,
  hex: _propTypes.default.string,
  hsl: _propTypes.default.object,
  hsv: _propTypes.default.object,
  rgb: _propTypes.default.object,
  onChange: _propTypes.default.func,
  disableAlpha: _propTypes.default.bool,
  readOnly: _propTypes.default.bool,
  renderers: _propTypes.default.func,
  onUnset: _propTypes.default.func
};
ColorPicker.defaultProps = {
  disableAlpha: false
};

var _default = (0, _common.ColorWrap)(ColorPicker);

exports.default = _default;
