// sanity-plugin-google-analytics/build/components/AuthProvider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactDom = require("react-dom");

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// dont wait for auth twice, even after unmounts
let isLoaded = false;

const initGoogleAPI = initCallback => {
  // Check that the google api is not initialized before
  if (typeof window !== 'undefined' && typeof gapi === 'undefined') {
    console.log('Init google api once') // eslint-disable-next-line max-params
    ;

    (function (w, d, s, g, js, fjs) {
      g = w.gapi || (w.gapi = {});
      g.analytics = {
        q: [],

        ready(cb) {
          this.q.push(cb);
          initCallback(g);
        }

      };
      js = d.createElement(s);
      fjs = d.getElementsByTagName(s)[0];
      js.src = 'https://apis.google.com/js/platform.js';
      fjs.parentNode.insertBefore(js, fjs);

      js.onload = function () {
        g.load('analytics');
      };
    })(window, document, 'script');
  }
}; // wait for auth to display children


class AuthProvider extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      ready: false
    });

    _defineProperty(this, "init", gapi => {
      const doAuth = () => {
        const authObj = this.props.accessToken ? {
          serverAuth: {
            access_token: this.props.accessToken
          }
        } : {
          clientid: this.props.clientId
        };
        gapi.analytics.auth && gapi.analytics.auth.authorize(_objectSpread({}, authObj, {
          container: this.authButtonNode
        }));
      };

      gapi.analytics.ready(a => {
        if (isLoaded) {
          this.setState({
            ready: true
          }); // if (this.props.onLoggedIn) {
          //   this.props.isLoggedIn()
          // }

          return;
        }

        const authResponse = gapi.analytics.auth.getAuthResponse();

        if (!authResponse) {
          gapi.analytics.auth.on("success", response => {
            console.log('Provider says logged in');
            this.setState({
              ready: true
            });
          });
        } else {
          console.log('Provider says logged in');
          this.setState({
            ready: true
          });
        }

        doAuth();
      });
    });
  }

  componentDidMount() {
    initGoogleAPI(this.init);
  }

  render() {
    return _react.default.createElement("div", {
      className: "sanity-plugin-google-analytics--provider"
    }, !this.props.clientId && _react.default.createElement("p", null, "Please provide clientId"), this.props.clientId && _react.default.createElement("div", {
      ref: node => this.authButtonNode = node
    }), this.state.ready && this.props.clientId && this.props.children);
  }

}

AuthProvider.propTypes = {
  clientId: _propTypes.default.string,
  accessToken: _propTypes.default.string
};
var _default = AuthProvider;
exports.default = _default;
