// sanity-plugin-google-analytics/build/components/CoreWidget.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/fieldsets/default"));

var _reactDebounceRender = _interopRequireDefault(require("react-debounce-render"));

var _spinner = _interopRequireDefault(require("part:@sanity/components/loading/spinner"));

var _configGoogleAnalyticsPlugin = _interopRequireDefault(require("config:google-analytics-plugin"));

var _GoogleMaterialDataChart = _interopRequireDefault(require("./GoogleMaterialDataChart"));

var _GoogleProvider = _interopRequireDefault(require("./GoogleProvider"));

var _Loading = _interopRequireDefault(require("./Loading"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const initGoogleAPI = () => {
  // Check that the google api is not initialized before
  if (typeof window !== 'undefined' && typeof gapi === 'undefined') {
    // eslint-disable-next-line max-params
    ;

    (function (w, d, s, g, js, fjs) {
      g = w.gapi || (w.gapi = {});
      g.analytics = {
        q: [],

        ready(cb) {
          this.q.push(cb);
        }

      };
      js = d.createElement(s);
      fjs = d.getElementsByTagName(s)[0];
      js.src = 'https://apis.google.com/js/platform.js';
      fjs.parentNode.insertBefore(js, fjs);

      js.onload = function () {
        g.load('analytics');
      };
    })(window, document, 'script');
  }
};

class CoreWidget extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "_interval", undefined);

    _defineProperty(this, "state", {
      hasGapi: false
    });

    _defineProperty(this, "init", () => {
      if (typeof window !== 'undefined' && window.gapi) {
        this.setState({
          hasGapi: true
        });
        clearInterval(this._interval);
      }
    });
  }

  componentDidMount() {
    initGoogleAPI();
    this._interval = setInterval(this.init, 100);
  }

  render() {
    const _this$props = this.props,
          type = _this$props.type,
          level = _this$props.level,
          clientId = _this$props.clientId,
          views = _this$props.views,
          children = _this$props.children,
          config = _this$props.config;
    const hasGapi = this.state.hasGapi;

    if (!_configGoogleAnalyticsPlugin.default) {
      return _react.default.createElement("p", null, "Please add ", _react.default.createElement("code", null, "google-analytics-plugin.json"), " to your config folder");
    }

    if (!hasGapi) {
      return _react.default.createElement(_Loading.default, null);
    }

    return _react.default.createElement(_GoogleProvider.default, {
      clientId: _configGoogleAnalyticsPlugin.default.clientId
    }, children || _react.default.createElement(_GoogleMaterialDataChart.default, _extends({}, this.props, {
      views: views || _configGoogleAnalyticsPlugin.default.views
    })));
  }

}

var _default = CoreWidget;
exports.default = _default;
