// sanity-plugin-google-analytics/build/components/DashboardWidget.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _CoreWidget = _interopRequireDefault(require("./CoreWidget"));

var _widgetContainer = _interopRequireDefault(require("part:@sanity/dashboard/widget-container"));

var _DashboardWidget = _interopRequireDefault(require("./DashboardWidget.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class DashboardWidget extends _react.default.Component {
  render() {
    const _this$props = this.props,
          gaConfig = _this$props.gaConfig,
          title = _this$props.title;
    return _react.default.createElement("div", {
      className: _DashboardWidget.default.container
    }, gaConfig ? _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("header", {
      className: _DashboardWidget.default.header
    }, _react.default.createElement("h3", {
      className: _DashboardWidget.default.title
    }, title || 'No title set')), _react.default.createElement("div", {
      className: _DashboardWidget.default.chart
    }, _react.default.createElement(_CoreWidget.default, {
      config: gaConfig
    }))) : _react.default.createElement("p", null, "Use ", _react.default.createElement("code", null, "gaConfig"), " to config your google analytics widget"));
  }

}

var _default = {
  name: 'google-analytics',
  component: DashboardWidget
};
exports.default = _default;
