// sanity-plugin-google-analytics/build/components/FormBuilderWidget.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/fieldsets/default"));

var _formBuilder = require("part:@sanity/form-builder");

var _CoreWidget = _interopRequireDefault(require("./CoreWidget"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

class FormBuilderWidget extends _react.default.Component {
  render() {
    const _this$props = this.props,
          type = _this$props.type,
          level = _this$props.level,
          document = _this$props.document;
    const _type$options = type.options,
          options = _type$options === void 0 ? {} : _type$options;
    return _react.default.createElement(_default2.default, {
      level: level,
      legend: type.title,
      description: type.description
    }, options && options.gaConfig ? _react.default.createElement(_CoreWidget.default, _extends({}, options, {
      config: options.gaConfig(document)
    })) : _react.default.createElement("p", null, "Use ", _react.default.createElement("code", null, "gaConfig"), " on ", _react.default.createElement("options", null, "options"), " to config your google analytics widget"));
  }

}

var _default = (0, _formBuilder.withDocument)(FormBuilderWidget);

exports.default = _default;
