// sanity-plugin-google-analytics/build/components/GoogleProvider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class GoogleProvider extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      ready: false,
      loggedIn: false
    });

    _defineProperty(this, "authButtonNode", _react.default.createRef());

    _defineProperty(this, "init", () => {
      if (typeof gapi === 'undefined') {
        return;
      }

      const doAuth = () => {
        const authObj = {
          clientid: this.props.clientId
        };
        gapi.analytics.auth && gapi.analytics.auth.authorize(_objectSpread({}, authObj, {
          container: this.authButtonNode.current
        }));
      };

      gapi.analytics.ready(a => {
        const authResponse = gapi.analytics.auth.getAuthResponse();

        if (!authResponse) {
          gapi.analytics.auth.on("success", response => {
            this.setState({
              ready: true,
              loggedIn: true
            });
          });
        } else {
          this.setState({
            ready: true
          });
        }

        doAuth();
      });
    });
  }

  componentDidMount() {
    this.init();
  }

  render() {
    const loggedIn = this.state.loggedIn;
    const hiddenStyles = {
      opacity: 0,
      height: 0,
      overflow: 'hidden'
    };
    return _react.default.createElement("div", {
      className: "sanity-plugin-google-analytics--provider"
    }, this.props.clientId && _react.default.createElement("div", {
      style: loggedIn ? hiddenStyles : {},
      ref: this.authButtonNode
    }), this.state.ready && this.props.children);
  }

}

GoogleProvider.propTypes = {
  clientId: _propTypes.default.string
};
var _default = GoogleProvider;
exports.default = _default;
